[FrameworkBundle] [Translation] Update XLIFF source tags with new translation:update-xliff-sources command

squrious opened the pull request:
| Q             | A          |
|---------------|------------|
| Branch?       | 7.1        |
| Bug fix?      | no         |
| New feature?  | yes        |
| Deprecations? | no         |
| Issues        | Fix #33041 |
| License       | MIT        |

## Context

Currently, XLIFF files generated with the `translation:extract` command use the translation key to create the `<source>` tag:

```xml
<trans-unit id="hVNhFLX" resname="navbar.home">
  <source>navbar.home</source>
  <target>__navbar.home</target>
</trans-unit>
```

But usually this `<source>` tag contains the default locale text to be translated ([XLIFF 1.2](http://docs.oasis-open.org/xliff/v1.2/os/xliff-core.html#source), [XLIFF 2.0](http://docs.oasis-open.org/xliff/xliff-core/v2.0/xliff-core-v2.0.html#source)).

Also, the `<source>` tag is already used in `XliffFileLoader` to populate the catalogue metadata ([but only for v1](https://github.com/symfony/symfony/blob/d621936e8c638247de4c5d1cd75cc5ff55b84c3e/src/Symfony/Component/Translation/Loader/XliffFileLoader.php#L127)).

## Proposed solution

This PR adds three features:
- use the `source` metadata (if available) to create the `source` tag in `XliffFileDumper`
- use the `source` tag to populate the `source` metadata in `XliffFileLoader` for XLIFF 2.0
- add a `translation:update-xliff-sources` command to update the `source` tag in XLIFF files

## Result

Given the following entries in XLIFF files:

```xml
<!-- translations/messages.en.xlf -->

<trans-unit id="hVNhFLX" resname="navbar.home">
  <source>navbar.home</source>
  <target>Home</target>
</trans-unit>
```

```xml
<!-- translations/messages.fr.xlf -->

<trans-unit id="hVNhFLX" resname="navbar.home">
  <source>navbar.home</source>
  <target>__navbar.home</target>
</trans-unit>
```

Running

```bash
bin/console translation:update-xliff-sources
```

Will produce updates the files with:

```xml
<!-- translations/messages.en.xlf -->

<trans-unit id="hVNhFLX" resname="navbar.home">
  <source>Home</source>
  <target>Home</target>
</trans-unit>
```

```xml
<!-- translations/messages.fr.xlf -->

<trans-unit id="hVNhFLX" resname="navbar.home">
  <source>Home</source>
  <target>__navbar.home</target>
</trans-unit>
```

The command has options to specify locales, domains, and paths:

```text
Description:
  Update source tags with default locale targets in XLIFF files.

Usage:
  translation:update-xliff-sources [options] [--] [<paths>...]

Arguments:
  paths                    Paths to look for translations.

Options:
      --format=FORMAT      Override the default output format. [default: "xlf12"]
      --domains[=DOMAINS]  Specify the domain to update. (multiple values allowed)
      --locales=LOCALES    Specify the locale to update. (multiple values allowed)
  -h, --help               Display help for the given command. When no command is given display help for the list command
  -q, --quiet              Do not output any message
  -V, --version            Display this application version
      --ansi|--no-ansi     Force (or disable --no-ansi) ANSI output
  -n, --no-interaction     Do not ask any interactive question
  -e, --env=ENV            The Environment name. [default: "dev"]
      --no-debug           Switch off debug mode.
      --profile            Enables profiling (requires debug).
  -v|vv|vvv, --verbose     Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
  The translation:update-xliff-sources command updates the source tags in XLIFF files with the default locale translation if available.
  
  You can specify directories to update in the command arguments: 
  
    php bin/console translation:update-xliff-sources path/to/dir path/to/another/dir
  
  To restrict the updates to one or more locales, including the default locale itself, use the --locales option:
  
    php bin/console translation:update-xliff-sources --locales en --locales fr
  
  You can specify one or more domains to target with the --domains option. By default, all available domains for the targeted locales are used.
  
    php bin/console translation:update-xliff-sources --domains messages
  
```

squrious committed:
Add --update-source option in translation:extract

stof commented:
Copying the default locale in the source tags of files for other locales should probably be done in a separate command than `translation:extract` as this is a different kind of change. This action would be doable for all keys that exist in the translation files, not only for keys that can be extracted from usages.

squrious committed:
Move to a dedicated command in the Translation component

squrious commented:
I created a dedicated command in the Translation component

welcoMattic commented on src/Symfony/Component/Translation/Command/XliffUpdateSourcesCommand.php:
```diff
@@ -0,0 +1,203 @@
+            $defaultLocaleCatalogue = $translatorBag->getCatalogue($this->defaultLocale);
+
+            if (!$defaultLocaleCatalogue instanceof MetadataAwareInterface) {
+                $io->error(sprintf('The default locale catalogue must implement "%s" to be used in this tool.', MetadataAwareInterface::class));
```
```suggestion
                $io->error(sprintf('The default locale catalogue must implement "%s" to be used in by this command.', MetadataAwareInterface::class));
```

welcoMattic commented on src/Symfony/Component/Translation/Command/XliffUpdateSourcesCommand.php:
```diff
@@ -0,0 +1,203 @@
+                $currentCatalogue = $translatorBag->getCatalogue($locale);
+
+                if (!$currentCatalogue instanceof MessageCatalogue) {
+                    $io->warning(sprintf('The catalogue for locale "%s" must be an instance of "%s" to be used in this tool.', $locale, MessageCatalogue::class));
```
```suggestion
                    $io->warning(sprintf('The catalogue for locale "%s" must be an instance of "%s" to be used by this command.', $locale, MessageCatalogue::class));
```

welcoMattic reviewed (commented):
Thanks @squrious for this great work! I have 2 small comments about wording.
Otherwise your PR looks good to me 👍 

Can you rebase your branch please?

OskarStark commented on src/Symfony/Component/Translation/Command/XliffUpdateSourcesCommand.php:
```diff
@@ -0,0 +1,203 @@
+                new InputOption('locales', null, InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY, 'Specify the locale to update.', $this->enabledLocales),
+            ])
+            ->setHelp(<<<EOF
+The <info>%command.name%</info> command updates the source tags in XLIFF files with the default locale translation if available.
```
```suggestion
The <info>%command.name%</info> command updates the source tags in XLIFF files with the default translation locale if available.
```
?

OskarStark commented on src/Symfony/Component/Translation/Command/XliffUpdateSourcesCommand.php:
```diff
@@ -0,0 +1,203 @@
+    {
+        $io = new SymfonyStyle($input, $output);
+
+        $format = $input->getOption('format');
+
+        if (!\array_key_exists($format, self::FORMATS)) {
+            $io->error(sprintf('Unknown format "%s". Available formats are: %s.', $format, implode(', ', array_map(fn ($f) => '"'.$f.'"', array_keys(self::FORMATS)))));
+
+            return self::INVALID;
+        }
```
shouldn't we move this to `initialize` method?

squrious replied:
Not sure about that, I can't return any code from the `initialize` method to indicate the option is invalid, so I would have to throw an exception instead. 

I used the same logic as in [`TranslationUpdateCommand`](https://github.com/symfony/symfony/blob/7.2/src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php#L138).

OskarStark commented on src/Symfony/Component/Translation/Tests/Command/XliffUpdateSourcesCommandTest.php:
```diff
@@ -0,0 +1,331 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Command;
```
namespace is wrong

squrious committed:
Fixes

squrious commented:
Hello, I updated the PR according to your comments (not sure about the use of  `initialize` to check command options though). 

Also moved the changelog entry in the translation component to 7.2. 

Low deps tests fail but I guess it's expected as the PR covers both framework bundle and translation component.

fabpot commented:
> Low deps tests fail but I guess it's expected as the PR covers both framework bundle and translation component.

To fix the low deps tests, you need to bump the required version of the component in the bundle.
